packages/reactive-element/src/reactive-element.ts

```diff
@@ -1257,7 +1265,25 @@ export abstract class ReactiveElement
       ).getPropertyOptions(name);
       const hasChanged = options.hasChanged ?? notEqual;
       const newValue = this[name as keyof this];
-      if (hasChanged(newValue, oldValue)) {
+      let changed = hasChanged(newValue, oldValue);
+      // When there is no change, check a corner case that can occur when
+      // 1. there's a defaultValue which was not reflected
+      // 2. the property is subsequently set to the defaultValue.
+      // For example, `prop: {defaultValue: 'foo', reflect: true}`
+      // and el.prop = 'foo'. This should be considered a change if the
+      // attribute is not set.
```

Comment:
Nice comment, thanks!

```suggestion
      // attribute is not set because we will now reflect the property to the attribute.
```